File: bin/02_anno/anno_csv.R
### Date: 250828 anno_csv.R

# Load necessary libraries
suppressPackageStartupMessages({
    library(optparse)
    library(Seurat)
    library(tools) #file_path_sans_ext
    library(ggplot2)
})

# Define command-line options
option_list <- list(
    make_option("--input_rds",
                            type="character",
                            default="/data/work/Single-Cell-Pipeline/output/dataget/peanut_merge/H1314.hr.rds",
                            help="Path to the input rds file. [default: %default]"),
    make_option("--input_csv",
                            type="character",
                            default="/data/work/Single-Cell-Pipeline/Anno/input/anno_H1314.csv",
                            help="Path to the input CSV file. [default: %default]"),
    make_option("--reduction_key",
                            type="character",
                            default="umap",
                            help="The reduction key of single cell data")
)

opt_parser <- OptionParser(option_list = option_list)
opts <- parse_args(opt_parser)
input_rds <- opts$input_rds
input_csv <- opts$input_csv
reduction_key <- opts$reduction_key

# Read the rds file
seurat_obj <- readRDS(input_rds)
print(seurat_obj)

# Read the mapping CSV file (both columns as strings)
mapping_df <- read.csv(input_csv, stringsAsFactors = FALSE)
old_col <- names(mapping_df)[1]
new_col <- names(mapping_df)[2]

# Warn if new annotation column exists; note it will be overwritten.
if(new_col %in% colnames(seurat_obj@meta.data)){
    message(sprintf("Warning: Column '%s' already exists in meta data, it will be overwritten.", new_col))
}

# Print unique values of the old column in the metadata
message(" --- Unique values in the '", old_col, "' column ---")
print(unique(seurat_obj@meta.data[[old_col]]))

# Create a mapping vector: names are old values, elements are new values.
map_dict <- mapping_df[[2]]
names(map_dict) <- mapping_df[[1]]

# Apply mapping to create a new metadata column.
# (Ensure the new values are character strings.)
seurat_obj@meta.data[[new_col]] <- as.character(map_dict[seurat_obj@meta.data[[old_col]]])

# Prepare output filenames
file_name <- basename(input_rds)
base_name <- file_path_sans_ext(file_name)
output_rds <- paste0(base_name, "_anno.rds")
output_pdf <- paste0(base_name, "_anno.pdf")

message("Output rds file: ", output_rds)
saveRDS(seurat_obj, file = output_rds)

# Create UMAP plots colored by the old and new annotation columns.
# Assumes that the UMAP reduction is present as 'umap'
pdf(output_pdf)
DimPlot(seurat_obj, reduction = reduction_key, group.by = old_col, label = TRUE) + ggtitle(old_col)
DimPlot(seurat_obj, reduction = reduction_key, group.by = new_col, label = TRUE) + ggtitle(new_col)
dev.off()

File: bin/02_anno/anno_csv.py
### 2500828 anno_csv.py
### 为h5ad的细胞添加新的注释，input_csv的第一列为h5ad原来的列和值，第二列为新建的列和值
### 确保输入的h5ad的.obsm是X_umap=umap

import pandas as pd
import scanpy as sc
import os
import argparse

parser = argparse.ArgumentParser(description="Process input file paths.")
parser.add_argument(
    "--input_h5ad",
    type=str,
    default="/data/work/Single-Cell-Pipeline/output/dataget/H1314/H1314.h5ad",
    help="Path to the input h5ad file. (default: %(default)s)"
)
parser.add_argument(
    "--input_csv",
    type=str,
    default="/data/work/Single-Cell-Pipeline/Anno/input/anno_H1314.csv",
    help="Path to the input CSV file. (default: %(default)s)"
)
parser.add_argument(
    "--reduction_key",
    type=str,
    default="umap",
    help="The reduction key of single cell data"
)
args = parser.parse_args()
input_h5ad = args.input_h5ad
input_csv = args.input_csv
reduction_key = args.reduction_key

# load anndata object
adata = sc.read_h5ad(input_h5ad)
print(adata)
# read mapping csv
mapping_df = pd.read_csv(input_csv)
mapping_df = mapping_df.astype(str)
# construct dictionary between new name and old name
rename_dict = dict(zip(mapping_df[mapping_df.columns[0]].astype(str),
                       mapping_df[mapping_df.columns[1]].astype(str)))
# check the new name wethere repeat
if mapping_df.columns[1] in adata.obs.columns:
    print(f"Warning: {mapping_df.columns[1]} already exists in adata.obs, it will be overwritten.")
# check all valuse of old column
print("--- old column included unique values ---")
print(adata.obs[mapping_df.columns[0]].unique())
# adata.obs[mapping_df.columns[1]] = adata.obs[mapping_df.columns[0]].cat.rename_categories(rename_dict)
adata.obs[mapping_df.columns[1]] = (adata.obs[mapping_df.columns[0]].map(mapping_df.set_index(mapping_df.columns[0])[mapping_df.columns[1]]))
file_name = os.path.basename(input_h5ad)
base_name = os.path.splitext(file_name)[0] # 提取文件名（不包括扩展名）
output_h5ad = base_name + "_anno.h5ad" # 构造输出文件名
output_umap = base_name + "_anno.pdf"
print(f"Output h5ad file: {output_h5ad}")
adata.write_h5ad(output_h5ad, compression='gzip')

from matplotlib.backends.backend_pdf import PdfPages
import matplotlib.pyplot as plt
with PdfPages(output_umap) as pdf:
    # sc.pl.umap默认的.obsm是X_umap
    sc.pl.scatter(adata, basis=reduction_key, color=[mapping_df.columns[0],mapping_df.columns[1]], legend_loc="on data") #sc.pl.embedding(adata, basis='umap_test') 
    plt.savefig(pdf, format='pdf', dpi=100, bbox_inches='tight')
    plt.close()